File: frontend/src/api/upgrade.js
import { request } from './client.js'
import { getStoredUserId } from '../lib/session'

export async function submitUpgradeFeedback({ upgraded, feedback, willing_to_pay }) {
  const user_id = getStoredUserId()
  if (!user_id) throw new Error('Not authenticated')

  return request('/api/upgrade', {
    method: 'POST',
    auth: true,
    body: JSON.stringify({
      user_id,
      upgraded: Boolean(upgraded),
      feedback: feedback || '',
      willing_to_pay: willing_to_pay || '',
    }),
  })
}

export async function getUpgradeFeedback() {
  const user_id = getStoredUserId()
  if (!user_id) throw new Error('Not authenticated')
  return request(`/api/upgrade?user_id=${encodeURIComponent(user_id)}`, { auth: true })
}

File: go-service/internal/models/upgrade.go
package models

// UpsertUpgradeRequest is the payload for POST /api/upgrade.
type UpsertUpgradeRequest struct {
	UserID        string  `json:"user_id" binding:"required"`
	Upgraded      bool    `json:"upgraded"`
	Feedback      *string `json:"feedback"`
	WillingToPay  *string `json:"willing_to_pay"`
}

// UpgradeRecord is the response returned for GET and POST /api/upgrade.
type UpgradeRecord struct {
	UserID        string  `json:"user_id"`
	Upgraded      bool    `json:"upgraded"`
	Feedback      *string `json:"feedback"`
	WillingToPay  *string `json:"willing_to_pay"`
}
